File: ios/CommonWeek/FamilyPlanningDemo.swift
import Foundation

/// In-memory previews use the same family features across sheets and weeks during one launch.
@MainActor
final class FamilyPlanningDemo {
    static let shared = FamilyPlanningDemo()
    private var children = [ChildProfile(id: "demo-child", name: "Miriam", color: "#688173", calendarPreferenceIds: ["calendar-family"])]
    private var adultCalendars: [String: [String]] = [:]
    private var routines: [TaskRoutine] = []
    private var templates: [WeekTemplate] = []
    private var applied = Set<String>()
    private var reviews: [String: WeeklyReview] = [:]
    private var weeks: [String: WeeklyPlannerData] = [:]
    private var generated = Set<String>()

    func capture(_ planner: WeeklyPlannerData?) {
        if let planner, planner.isDemo { weeks[planner.weekStart] = planner }
    }

    func planner(weekStart: String, capturing existing: WeeklyPlannerData? = nil) -> WeeklyPlannerData {
        capture(existing)
        var value = weeks[weekStart] ?? PreviewData.planner(weekStart: weekStart)
        for routine in routines where routine.active {
            for occurrence in occurrences(routine, weekStart: weekStart) {
                let identity = "\(routine.id):\(occurrence ?? weekStart)"
                guard !generated.contains(identity) else { continue }
                generated.insert(identity)
                let item = PlanningItem(id: identity, planningDate: occurrence, weekStartDate: weekStart, type: .task, text: routine.text, isCompleted: false, sortOrder: 0, createdBy: PreviewData.user.userId, createdByName: "Jim", updatedAt: WeekDate.iso8601.string(from: Date()), saveState: "saved", reminder: nil, childId: routine.childId, routineId: routine.id, routineOccurrenceDate: occurrence ?? weekStart)
                insert(item, into: &value)
            }
        }
        for index in value.days.indices {
            value.days[index].events = value.days[index].events.map { event in
                var event = event
                event.assignedAdultUserIds = value.members.filter { member in
                    adultCalendars[member.userId]?.contains(event.calendarPreferenceId ?? event.calendarId) ?? (event.sourceUserId == member.userId)
                }.map(\.userId)
                return event
            }
        }
        weeks[weekStart] = value
        return value
    }

    func data(weekStart: String) -> FamilyPlanningData {
        let review = reviews[weekStart] ?? .init(weekStart: weekStart, priorities: "Keep Saturday afternoon open", meals: "Dinner at home Monday through Thursday", logistics: "Confirm camp pickup plans", revision: 0, reviewedBy: [])
        let planner = weeks[weekStart] ?? PreviewData.planner(weekStart: weekStart)
        let adults = planner.members.map { member in
            AdultCalendarAssignment(userId: member.userId, displayName: member.displayName, calendarPreferenceIds: adultCalendars[member.userId] ?? CalendarEventFilter.calendars(in: planner).filter { $0.sourceUserId == member.userId }.map(\.id))
        }
        return FamilyPlanningData(weekStart: weekStart, currentUserId: PreviewData.user.userId, canEdit: true, children: children, routines: routines, templates: templates.map { .init(id: $0.id, name: $0.name, items: $0.items, appliedToWeek: applied.contains("\($0.id):\(weekStart)")) }, review: review, openTasks: weeks.values.filter { $0.weekStart < weekStart }.flatMap { $0.weeklyItems + $0.days.flatMap(\.items) }.filter { $0.type == .task && !$0.isCompleted }, adults: adults)
    }

    func apply(_ mutation: FamilyPlanningMutation) throws -> FamilyPlanningData {
        let week = mutation.weekStart
        var review = data(weekStart: week).review
        switch mutation.action {
        case "saveAdultCalendars":
            if let userId = mutation.userId, let ids = mutation.calendarPreferenceIds { adultCalendars[userId] = ids }
        case "saveChild":
            if let child = mutation.child { children.removeAll { $0.id == child.id }; children.append(child) }
        case "deleteChild":
            children.removeAll { $0.id == mutation.id }
            for key in Array(weeks.keys) {
                transformItems(in: key) { item in var item = item; if item.childId == mutation.id { item.childId = nil }; return item }
            }
        case "saveRoutine", "deleteRoutine":
            let id = mutation.routine?.id ?? mutation.id
            routines.removeAll { $0.id == id }
            if let routine = mutation.routine, mutation.action == "saveRoutine" { routines.append(routine) }
            for key in Array(weeks.keys) {
                transformItems(in: key) { item in
                    guard item.routineId == id, !item.isCompleted, (item.routineOccurrenceDate ?? item.weekStartDate) >= WeekDate.today(timeZoneIdentifier: "America/New_York") else { return item }
                    generated.remove("\(item.routineId ?? ""):\(item.routineOccurrenceDate ?? item.weekStartDate)")
                    return nil
                }
            }
            if let sourceId = mutation.sourceItemId, let routine = mutation.routine {
                for key in Array(weeks.keys) {
                    transformItems(in: key) { item in
                        guard item.id == sourceId else { return item }
                        var updated = item
                        updated.routineId = routine.id
                        updated.routineOccurrenceDate = item.planningDate ?? item.weekStartDate
                        updated.childId = routine.childId
                        generated.insert("\(routine.id):\(updated.routineOccurrenceDate!)")
                        return updated
                    }
                }
            }
        case "saveTemplate":
            let snapshot = planner(weekStart: week)
            let items = (snapshot.weeklyItems + snapshot.days.flatMap(\.items)).filter { $0.routineId == nil }.map {
                WeekTemplate.Item(dayOffset: $0.planningDate.map { WeekDate.daysBetween(week, $0) }, type: $0.type, text: $0.text, childId: $0.childId)
            }
            guard !items.isEmpty else { throw APIError.server("Add a plan or one-off task to save as a template.") }
            let id = mutation.id ?? UUID().uuidString
            if !templates.contains(where: { $0.id == id }) {
                templates.append(.init(id: id, name: mutation.name ?? "Week template", items: items, appliedToWeek: false))
                applied.insert("\(id):\(week)")
            }
        case "deleteTemplate": templates.removeAll { $0.id == mutation.id }
        case "applyTemplate":
            guard let template = templates.first(where: { $0.id == mutation.id }) else { throw APIError.invalidResponse }
            let key = "\(template.id):\(week)"
            if !applied.contains(key) {
                var value = planner(weekStart: week)
                for (index, item) in template.items.enumerated() {
                    insert(.init(id: "\(key):\(index)", planningDate: item.dayOffset.map { WeekDate.addDays($0, to: week) }, weekStartDate: week, type: item.type, text: item.text, isCompleted: false, sortOrder: index, createdBy: PreviewData.user.userId, createdByName: "Jim", updatedAt: WeekDate.iso8601.string(from: Date()), saveState: "saved", reminder: nil, childId: item.childId), into: &value)
                }
                weeks[week] = value; applied.insert(key)
            }
        case "saveReview":
            guard mutation.revision == review.revision else { throw APIError.server("The shared plan changed. Reload before saving your notes.") }
            review = .init(weekStart: week, priorities: mutation.priorities ?? "", meals: mutation.meals ?? "", logistics: mutation.logistics ?? "", revision: review.revision + 1, reviewedBy: [])
            reviews[week] = review
        case "markReviewed":
            guard mutation.revision == review.revision else { throw APIError.server("The shared plan changed. Read the latest plan before marking it reviewed.") }
            reviews[week] = .init(weekStart: week, priorities: review.priorities, meals: review.meals, logistics: review.logistics, revision: review.revision, reviewedBy: mutation.reviewed == true ? [.init(userId: PreviewData.user.userId, displayName: "Jim", reviewedAt: WeekDate.iso8601.string(from: Date()))] : [])
        default: throw APIError.invalidResponse
        }
        return data(weekStart: week)
    }

    private func insert(_ item: PlanningItem, into planner: inout WeeklyPlannerData) {
        if let date = item.planningDate, let index = planner.days.firstIndex(where: { $0.date == date }) { planner.days[index].items.append(item) }
        else { planner.weeklyItems.append(item) }
    }

    private func transformItems(in week: String, transform: (PlanningItem) -> PlanningItem?) {
        guard var planner = weeks[week] else { return }
        planner.weeklyItems = planner.weeklyItems.compactMap(transform)
        for index in planner.days.indices { planner.days[index].items = planner.days[index].items.compactMap(transform) }
        weeks[week] = planner
    }

    private func occurrences(_ routine: TaskRoutine, weekStart: String) -> [String?] {
        if routine.frequency == "weekly", routine.weekdays.isEmpty {
            let distance = WeekDate.daysBetween(WeekDate.weekStart(for: routine.startsOn), weekStart) / 7
            return distance >= 0 && distance % max(1, routine.interval) == 0 && (routine.endsOn == nil || weekStart <= routine.endsOn!) ? [nil] : []
        }
        return (0..<7).compactMap { offset -> String? in
            let date = WeekDate.addDays(offset, to: weekStart)
            guard date >= routine.startsOn, routine.endsOn == nil || date <= routine.endsOn! else { return nil }
            if routine.frequency == "daily" {
                return WeekDate.daysBetween(routine.startsOn, date) % max(1, routine.interval) == 0
                    && (routine.weekdays.isEmpty || routine.weekdays.contains(offset)) ? date : nil
            }
            let weeks = WeekDate.daysBetween(WeekDate.weekStart(for: routine.startsOn), weekStart) / 7
            return weeks % max(1, routine.interval) == 0 && routine.weekdays.contains(offset) ? date : nil
        }.map(Optional.some)
    }
}

File: src/components/planner/family-planning-demo.ts
import { addDateDays, todayInTimeZone } from "@/lib/date";
import { routineOccurrencesForWeek } from "@/lib/family-planning";
import type { FamilyPlanningData, FamilyPlanningMutation, PlanningItem, TaskRoutine, WeeklyPlannerData, WeeklyReview } from "@/types/domain";

const storageKey = "week-of-us:family-planning-demo:v1";
interface DemoStore {
  adultCalendars?: Record<string, string[]>;
  children: FamilyPlanningData["children"];
  routines: FamilyPlanningData["routines"];
  templates: FamilyPlanningData["templates"];
  appliedWeeks: Record<string, string[]>;
  reviews: Record<string, WeeklyReview>;
  itemsByWeek: Record<string, PlanningItem[]>;
}
const blankStore = (): DemoStore => ({ children: [], routines: [], templates: [], appliedWeeks: {}, reviews: {}, itemsByWeek: {} });
function readStore(): DemoStore {
  try { const raw = localStorage.getItem(storageKey); return raw ? { ...blankStore(), ...JSON.parse(raw) } : blankStore(); } catch { return blankStore(); }
}
function writeStore(store: DemoStore) { try { localStorage.setItem(storageKey, JSON.stringify(store)); } catch { /* A private browser can still use the current demo session. */ } }
export function emptyFamilyPlanning(weekStart: string, currentUserId: string): FamilyPlanningData {
  return { weekStart, currentUserId, canEdit: true, adults: [], children: [], routines: [], templates: [], openTasks: [], review: { weekStart, priorities: "", meals: "", logistics: "", revision: 0, reviewedBy: [] } };
}
function newTask(data: WeeklyPlannerData, values: Partial<PlanningItem> & Pick<PlanningItem, "id" | "text" | "planningDate">): PlanningItem {
  return { weekStartDate: data.weekStart, type: "task", isCompleted: false, sortOrder: 0, createdBy: data.members[0]?.userId ?? "demo-user", createdByName: data.members[0]?.displayName ?? "You", updatedAt: new Date().toISOString(), saveState: "saved", ...values };
}
function materialize(data: WeeklyPlannerData, store: DemoStore, items: PlanningItem[]): PlanningItem[] {
  const next = [...items];
  for (const routine of store.routines) {
    for (const occurrence of routineOccurrencesForWeek(routine, data.weekStart)) {
      const id = `demo-routine-${routine.id}-${occurrence.occurrenceDate}`;
      if (!next.some((item) => item.id === id || (item.routineId === routine.id && item.routineOccurrenceDate === occurrence.occurrenceDate))) next.push(newTask(data, { id, text: routine.text, planningDate: occurrence.planningDate, childId: routine.childId, routineId: routine.id, routineOccurrenceDate: occurrence.occurrenceDate }));
    }
  }
  return next;
}
function familyFromStore(data: WeeklyPlannerData, currentUserId: string, store: DemoStore): FamilyPlanningData {
  const adults = data.members.map((member) => ({ userId: member.userId, displayName: member.displayName, calendarPreferenceIds: store.adultCalendars?.[member.userId] ?? data.visibleCalendars.filter((calendar) => calendar.sourceUserId === member.userId).map((calendar) => calendar.id) }));
  return { ...emptyFamilyPlanning(data.weekStart, currentUserId), adults, children: store.children, routines: store.routines, templates: store.templates.map((template) => ({ ...template, appliedToWeek: (store.appliedWeeks[template.id] ?? []).includes(data.weekStart) })), review: store.reviews[data.weekStart] ?? emptyFamilyPlanning(data.weekStart, currentUserId).review, openTasks: Object.entries(store.itemsByWeek).filter(([week]) => week < data.weekStart).flatMap(([, items]) => items).filter((item) => item.type === "task" && !item.isCompleted).slice(-100) };
}
export function loadDemoFamilyPlanning(data: WeeklyPlannerData, currentUserId: string): { family: FamilyPlanningData; items: PlanningItem[] } {
  const store = readStore();
  const initialItems = [...data.days.flatMap((day) => day.items), ...data.weeklyItems];
  const items = materialize(data, store, store.itemsByWeek[data.weekStart] ?? initialItems);
  store.itemsByWeek[data.weekStart] = items;
  writeStore(store);
  return { family: familyFromStore(data, currentUserId, store), items };
}
export function persistDemoPlanningItems(weekStart: string, items: PlanningItem[]) {
  const store = readStore();
  store.itemsByWeek[weekStart] = items.filter((item) => item.weekStartDate === weekStart);
  writeStore(store);
}
export function updateDemoPriorItem(item: PlanningItem) {
  const store = readStore();
  for (const week of Object.keys(store.itemsByWeek)) store.itemsByWeek[week] = store.itemsByWeek[week].filter((candidate) => candidate.id !== item.id);
  store.itemsByWeek[item.weekStartDate] = [...(store.itemsByWeek[item.weekStartDate] ?? []), item];
  writeStore(store);
}
export function mutateDemoFamilyPlanning(data: WeeklyPlannerData, currentUserId: string, currentUserName: string, mutation: FamilyPlanningMutation, currentItems: PlanningItem[]): { family: FamilyPlanningData; items: PlanningItem[]; error?: string } {
  const store = readStore();
  let items = [...currentItems];
  store.itemsByWeek[data.weekStart] = items;
  const review = store.reviews[data.weekStart] ?? emptyFamilyPlanning(data.weekStart, currentUserId).review;
  switch (mutation.action) {
    case "saveAdultCalendars":
      store.adultCalendars = { ...store.adultCalendars, [mutation.userId]: mutation.calendarPreferenceIds };
      break;
    case "saveChild": {
      const child = { ...mutation.child, id: mutation.child.id ?? crypto.randomUUID() };
      store.children = [...store.children.filter((candidate) => candidate.id !== child.id), child];
      break;
    }
    case "deleteChild":
      store.children = store.children.filter((child) => child.id !== mutation.id);
      store.routines = store.routines.map((routine) => routine.childId === mutation.id ? { ...routine, childId: null } : routine);
      store.templates = store.templates.map((template) => ({ ...template, items: template.items.map((item) => item.childId === mutation.id ? { ...item, childId: null } : item) }));
      for (const week of Object.keys(store.itemsByWeek)) store.itemsByWeek[week] = store.itemsByWeek[week].map((item) => item.childId === mutation.id ? { ...item, childId: null } : item);
      items = store.itemsByWeek[data.weekStart];
      break;
    case "saveRoutine":
    case "deleteRoutine": {
      const id = mutation.action === "saveRoutine" ? mutation.routine.id ?? crypto.randomUUID() : mutation.id;
      const routine: TaskRoutine | null = mutation.action === "saveRoutine" ? { ...mutation.routine, id, childId: mutation.routine.childId ?? null, endsOn: mutation.routine.endsOn ?? null } : null;
      store.routines = store.routines.filter((candidate) => candidate.id !== id);
      if (routine) store.routines.push(routine);
      const today = todayInTimeZone(data.household.timezone);
      for (const week of Object.keys(store.itemsByWeek)) {
        const valid = routine ? routineOccurrencesForWeek(routine, week) : [];
        store.itemsByWeek[week] = store.itemsByWeek[week].flatMap((item) => {
          if (item.routineId !== id || item.isCompleted || (item.routineOccurrenceDate ?? item.planningDate ?? item.weekStartDate) < today) return [item];
          return valid.some((occurrence) => occurrence.occurrenceDate === item.routineOccurrenceDate) && routine ? [{ ...item, text: routine.text, childId: routine.childId }] : [];
        });
      }
      if (mutation.action === "saveRoutine" && mutation.sourceItemId && routine) {
        const source = store.itemsByWeek[data.weekStart].find((item) => item.id === mutation.sourceItemId);
        const occurrence = routineOccurrencesForWeek(routine, data.weekStart).find((candidate) => candidate.planningDate === source?.planningDate);
        if (!source || !occurrence) return { family: familyFromStore(data, currentUserId, store), items, error: "Choose a repeat day and start date that include this task’s current placement." };
        store.itemsByWeek[data.weekStart] = store.itemsByWeek[data.weekStart].map((item) => item.id === source.id ? { ...item, text: routine.text, childId: routine.childId, routineId: routine.id, routineOccurrenceDate: occurrence.occurrenceDate } : item);
      }
      items = materialize(data, store, store.itemsByWeek[data.weekStart]);
      break;
    }
    case "saveTemplate": {
      const id = mutation.id ?? crypto.randomUUID();
      if (!store.templates.some((template) => template.id === id)) store.templates.push({ id, name: mutation.name, appliedToWeek: false, items: items.filter((item) => item.weekStartDate === data.weekStart && !item.routineId).map((item) => ({ type: item.type, text: item.text, childId: item.childId ?? null, dayOffset: item.planningDate ? Math.round((Date.parse(item.planningDate) - Date.parse(data.weekStart)) / 86_400_000) : null })) });
      break;
    }
    case "deleteTemplate": store.templates = store.templates.filter((template) => template.id !== mutation.id); break;
    case "applyTemplate": {
      const template = store.templates.find((candidate) => candidate.id === mutation.id);
      if (template && !(store.appliedWeeks[template.id] ?? []).includes(data.weekStart)) {
        items.push(...template.items.map((item, index) => newTask(data, { id: `demo-template-${template.id}-${data.weekStart}-${index}`, text: item.text, type: item.type, childId: item.childId, planningDate: item.dayOffset === null ? null : addDateDays(data.weekStart, item.dayOffset) })));
        store.appliedWeeks[template.id] = [...(store.appliedWeeks[template.id] ?? []), data.weekStart];
      }
      break;
    }
    case "assignChild": items = items.map((item) => item.id === mutation.itemId ? { ...item, childId: mutation.childId } : item); break;
    case "saveReview":
      if (review.revision !== mutation.revision) return { family: familyFromStore(data, currentUserId, store), items, error: "The shared notes changed. Reopen the planner to review the latest notes; your draft is still here." };
      store.reviews[data.weekStart] = { ...review, priorities: mutation.priorities, meals: mutation.meals, logistics: mutation.logistics, revision: review.revision + 1, reviewedBy: [] };
      break;
    case "markReviewed":
      if (mutation.revision !== undefined && review.revision !== mutation.revision) return { family: familyFromStore(data, currentUserId, store), items, error: "The plan changed. Review the latest notes before marking the week reviewed." };
      store.reviews[data.weekStart] = { ...review, reviewedBy: [...review.reviewedBy.filter((member) => member.userId !== currentUserId), ...(mutation.reviewed ? [{ userId: currentUserId, displayName: currentUserName, reviewedAt: new Date().toISOString() }] : [])] };
      break;
  }
  store.itemsByWeek[data.weekStart] = items;
  writeStore(store);
  return { family: familyFromStore(data, currentUserId, store), items };
}
